AsyncNats/Messages/NatsMessageParser.cs

```diff
@@ -2,80 +2,407 @@
-                var message = parseMessage(_memoryPool, line, ref reader);
-                if (message == null)
+                else
                 {
-                    // Not enough information to parse the message
-                    reader.Rewind(reader.Consumed - previousPosition);
-                    break;
+                    inlined++;
                 }
-
-                outMessages[messageCount] = message;
-                messageCount++;
                
             }
 
             consumed = reader.Consumed;
             return messageCount;
         }
+
+        public NatsMsg? ParseMessage(in ReadOnlySpan<byte> line, ref SequenceReader<byte> reader, out bool inlined)
+        {
+            //the idea here is to walk the *line* backwards in sequence just once
+            //this should prevent memory fetching as it will usually fit into one cache line
+            //it also helps that parsing int is done from right to left
+            //we trust nats won't send malformed header lines
+
+            inlined = false;
+
+            //parse total size
+            var multiplier = 1;
+            var payloadSize = 0;
+            var pointer = line.Length - 1;
+
+            ref byte lineRef = ref MemoryMarshal.GetReference(line);
+            byte currentByte = Unsafe.Add(ref MemoryMarshal.GetReference(line), pointer);
+            do
+            {
+                payloadSize += (currentByte - '0') * multiplier;
+                multiplier *= 10;
+                pointer--;
+                currentByte = Unsafe.Add(ref lineRef, pointer);
+            } while (currentByte != ' ');
+
+            if (reader.Remaining < payloadSize + 2) return null;
+
+            var payloadSizeStart = pointer;
+            pointer--;
+
+            Span<int> splits = stackalloc int[3];
+            var splitCount = 0;
+
+            while (pointer > 3)
+            {
+                var ch = line[pointer];
+                if (ch == ' ')
+                {
+                    splits[splitCount] = pointer;
+                    splitCount++;
+
+                    if (splitCount > 2)
+                        throw new ProtocolViolationException($"Invalid message header {Encoding.UTF8.GetString(line)}");
+                }
+
+                pointer--;
+            }
+            //done with first line
+
+            
+
+            long sid = 0;
+            pointer = (splitCount == 1) ? payloadSizeStart - 1 : splits[0] - 1;
+            //sid            
+            multiplier = 1;
+            currentByte = Unsafe.Add(ref lineRef, pointer);
+            do
+            {
+                sid += (currentByte - '0') * multiplier;
+                multiplier *= 10;
+                pointer--;
+                currentByte = Unsafe.Add(ref lineRef, pointer);
+            } while (currentByte != ' ');
+
+            if (_inlineSubscriptions.TryGetValue(sid, out var inlineSubscription))
```

Comment:
There's a big chance (when looking at reference code) that having a boolean "hasInlineSubscriptions" can have a positive effect as TryGetValue does a lot of work for (event) empty dictionaries. See: https://github.com/microsoft/referencesource/blob/master/mscorlib/system/collections/Concurrent/ConcurrentDictionary.cs#L498

Reply:
You are right. The issue here is that the dictionary is mutable.  There's probably room for improvement by using an immutable dictionary and swapping references. I feel it can be addressed in a second pass.